### ISSUE TITLE
Help with glob syntax (Next.js special routes related)

### ISSUE THREAD

**Issue by milos-sikic-nimbus-tech** [DukeLuo/eslint-plugin-check-file#23]
Hi there!

I've been trying to make a glob pattern that would match what this regex matches for a while:

```regex
^(?:\[[a-z]+(?:-[a-z]+)*]|(?:[a-z]+(?:-[a-z]+)*))$
```

But unfortunately, I can't figure it out. 

I have seen #20 and I agree with you that it should not be part of the library, but I would like to ask for your help for a glob that would match what's needed, if it's possible. If not, would you consider supporting regexes?

### Should pass

some
some-string
[some-string]
[some]

### Should not pass

[some-string
some-string]
[some]-string

Thank you for the tool. I would happily submit a PR that documents these custom globs for Next.js if we can make some.

**Comment by DukeLuo** [comment]
Hi @milos-sikic-nimbus-tech, I have created a glob for [your case](https://globster.xyz/?q=**%2F%40(%2B(%5Ba-z%5D)*(-%2B(%5Ba-z%5D))%7C%5C%5B%2B(%5Ba-z%5D)*(-%2B(%5Ba-z%5D))%5C%5D).js&f=%2Fmyapp%2Fsrc%2Fsome.js%2C%2Fmyapp%2Fsrc%2Fsome-string.js%2C%2Fmyapp%2Fsrc%2F%5Bsome-string%5D.js%2C%2Fmyapp%2Fsrc%2F%5Bsome%5D.js%2C%2Fmyapp%2Fsrc%2F%5Bsome-string.js%2C%2Fmyapp%2Fsrc%2Fsome-string%5D.js%2C%2Fmyapp%2Fsrc%2F%5Bsome%5D-string.js) `@(+([a-z])*(-+([a-z]))|\[+([a-z])*(-+([a-z]))\])`, it looks a bit complex. You can test it to see if it fits your needs.

This glob combines different patterns to match either a word without brackets or a word enclosed in square brackets, allowing hyphens within the words:
- `+([a-z])*(-+([a-z]))` is used to match a word
- `\[+([a-z])*(-+([a-z]))\]` is used to match a word enclosed in square brackets

**Comment by milos-sikic-nimbus-tech** [comment]
Hello, @DukeLuo, thank you very much for the prompt response!

Unfortunately, the pattern seems to not match anything when I use it like this.

```
...
      'check-file/folder-naming-convention': [
        'error',
        {
          'src/**/': '@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])',
          'app/**/': '@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])',
          'pages/**/': '@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])',
        },
      ],
...
```

Output:

```
$ ESLINT_USE_FLAT_CONFIG=true eslint -c eslint.config.mjs .

/projects/nextjs-template/src/app/[locale]/error.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/example-route/page.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/example-route/styles.module.scss.d.ts
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/layout.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/loader.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/loading.module.scss.d.ts
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/loading.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/locale.props.ts
  13:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/page.module.scss.d.ts
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/page.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/providers.tsx
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/app/[locale]/routes.ts
  1:1  error  The folder "app" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/assets/icons/index.ts
  1:1  error  The folder "assets" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/components/component-example/component-example.module.scss.d.ts
  1:1  error  The folder "components" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/components/component-example/component-example.tsx
  1:1  error  The folder "components" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/translations/client-translations.ts
  1:1  error  The folder "translations" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/translations/server-translations.ts
  1:1  error  The folder "translations" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

/projects/nextjs-template/src/translations/settings.ts
  1:1  error  The folder "translations" does not match the "@(+([a-z])*(-+([a-z]))|[+([a-z])*(-+([a-z]))])" pattern  check-file/folder-naming-convention

✖ 18 problems (18 errors, 0 warnings)
```

I also tried with using just the `app` matcher inside of `eslint.config.mjs`, but it just failed to match the `[locale]` folder then.

Let me know if I can help out somehow, and again, thank you for your time!

**Comment by milos-sikic-nimbus-tech** [comment]
Thank you very much, @DukeLuo, that works for me, too. I was sure that I tried with only the first rule, but it seems I didn't.

Would https://github.com/DukeLuo/eslint-plugin-check-file/blob/main/docs/rules/folder-naming-convention.md be a good page to add a Next.js example? Or maybe a new page with various examples?

**Comment by DukeLuo** [comment]
> Thank you very much, @DukeLuo, that works for me, too. I was sure that I tried with only the first rule, but it seems I didn't.
> 
> Would https://github.com/DukeLuo/eslint-plugin-check-file/blob/main/docs/rules/folder-naming-convention.md be a good page to add a Next.js example? Or maybe a new page with various examples?

That's cool, glad to hear it. I think it's a good idea to include a Next.js example on the original page. If someone needs this glob, it's easier to find.

**Comment by milos-sikic-nimbus-tech** [comment]
Sorry @DukeLuo, but I'm not sure what page you're referring to when you say original. Just send me a link to the page where you'd like me to add the example and I'll send a PR!

**Comment by DukeLuo** [comment]
> Sorry @DukeLuo, but I'm not sure what page you're referring to when you say original. Just send me a link to the page where you'd like me to add the example and I'll send a PR!

Yes, I mean https://github.com/DukeLuo/eslint-plugin-check-file/blob/main/docs/rules/folder-naming-convention.md. Thanks for your PR!

**Comment by milos-sikic-nimbus-tech** [comment]
PR opened #27

**Comment by DukeLuo** [comment]
`NEXT_JS_APP_ROUTER_CASE` was released in v2.5.0.
